### PR TITLE
Support {smallmatrix}, {subarray}, and \substack

### PR DESCRIPTION
**Pull request by ronkok**
Support two new environments and a function. Addresses #1967.

Here's a screenshot that compares `{matrix}` to `{smallmatrix}` to `{subarray}` to `\substack`:
small

The same code from quicklatex.com looks like this:
latexSmall

### CODE REVIEW

**Comment by ronkok** [comment]
Screenshots and documentation is in. This should be ready to review.

I omitted `{subarray}` from the environment documentation, and mentioned just `\substack` instead. The AMS Users Guide does roughly the same thing.

In the macro for `\substack`, I had to prevent a `\\` in the argument from throwing a parse error. I have successfully done that, but I have a nagging suspicion that I could have done so in a more elegant manner. @edemaine if you have any suggestions, I would appreciate it.

**Comment by kevinbarabash** [comment]
@ronkok this is awesome.  The `\smallmatrix` output looks so much better than quicklatex.com.  `\substack` seems really useful.  Thanks for adding these commands.

File: src/environments/array.js [hunk 1]
```diff
@@ -401,7 +404,9 @@ const mathmlBuilder: MathMLBuilder<"array"> = function(group, options) {
 
     // The 0.16 and 0.09 values are found emprically. They produce an array
     // similar to LaTeX and in which content does not interfere with \hines.
-    const gap = 0.16 + group.arraystretch - 1 + (group.addJot ? 0.09 : 0);
+    const gap = (group.arraystretch === 0.5)
+        ? "0.1"  // {smallmatrix}, {subarray}
```
[the remark this answers is not in the document]
Comment (ronkok) [reply src/environments/array.js]: Fair point. It's now a number.

File: src/environments/array.js [hunk 2]
```diff
@@ -484,13 +491,18 @@ const mathmlBuilder: MathMLBuilder<"array"> = function(group, options) {
...
     }
 
-    if (menclose === "") {
-        return table;
-    } else {
-        const wrapper =  new mathMLTree.MathNode("menclose", [table]);
-        wrapper.setAttribute("notation", menclose.trim());
-        return wrapper;
+    if (menclose !== "") {
+        table =  new mathMLTree.MathNode("menclose", [table]);
```
[the remark this answers is not in the document]
Comment (ronkok) [reply src/environments/array.js]: Done.

File: src/environments/array.js [hunk 3]
```diff
@@ -484,13 +491,18 @@ const mathmlBuilder: MathMLBuilder<"array"> = function(group, options) {
...
+        table.setAttribute("notation", menclose.trim());
+    }
+
+    if (group.arraystretch && group.arraystretch < 1) {
+        // A small array. Wrap in scriptstyle so row gap is not too large.
+        table =  new mathMLTree.MathNode("mstyle", [table]);
```
[the remark this answers is not in the document]
Comment (ronkok) [reply src/environments/array.js]: Done.

File: src/environments/array.js [hunk 4]
```diff
@@ -656,6 +668,63 @@ defineEnvironment({
...
+    type: "array",
+    names: ["subarray"],
+    props: {
+        numArgs: 1,
+    },
+    handler(context, args) {
+        // Parsing of {subarray} is similar to {array}
+        const symNode = checkSymbolNodeType(args[0]);
+        const colalign: AnyParseNode[] =
+            symNode ? [args[0]] : assertNodeType(args[0], "ordgroup").body;
```
[the remark this answers is not in the document]
Comment (ronkok) [reply src/environments/array.js]: I believe you are correct.

This line of code is a direct copy from the `{array}` and `{darray}` environments. I want the same behavior for `{subarray}` and I did not feel a need to improve the coding style.

File: src/environments/array.js [hunk 5]
```diff
@@ -656,6 +668,63 @@ defineEnvironment({
...
+        const cols = colalign.map(function(nde) {
```
[the remark this answers is not in the document]
Comment (ronkok) [reply src/environments/array.js]: If I rewrite this here, I should also edit the `{array}` code. Is that what you want?

File: src/macros.js [hunk 6]
```diff
@@ -426,6 +426,28 @@ defineMacro("\\varPhi", "\\mathit{\\Phi}");
 defineMacro("\\varPsi", "\\mathit{\\Psi}");
 defineMacro("\\varOmega", "\\mathit{\\Omega}");
 
+//\newcommand{\substack}[1]{\subarray{c}#1\endsubarray}
+defineMacro("\\substack", function(context) {
+    // A simple expansion of the argument causes a Parse error at "\\".
+    // So recreate the original text of the argument.
```
[the remark this answers is not in the document]
Comment (ronkok) [reply src/macros.js]: It works for me, too. I've edited the file to have the nice one-liner macro.

**Comment by ronkok** [comment]
Okay, I've picked up the review comments and fixed the `{smallmatrix}` column separator distance. This should be ready to go.

File: src/environments/array.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply src/environments/array.js]: Since this was copied from `{array}` let's leave this as is.

**Pull request merged**